activerecord/lib/active_record/insert_all.rb

```diff
@@ -127,6 +141,34 @@ def verify_attributes(attributes)
         end
       end
 
+      class CommonAssociationExtender
+        def initialize(reflection)
+          @reflection = reflection
+        end
+
+        def keys
+          [@reflection.foreign_key]
+        end
+
+        def attributes(object)
+          { @reflection.foreign_key => object.id }
```

Comment:
`object.id` is not always correct for association primary key.

See also #40815, #41090.

Reply:
@kamipo Is there a way like `primary_key`? I am not aware of the best way I need to go other what we can use in this case and probably for a complex primary key.

My cleanup with extenders is result of previous needs with two checks on polymorphic and two extraction of reflection in `initialize` and for attributes. I try to create yet one extenders but could not find out a good way

Reply:
Also it looks like I need to use some approaches from associations to avoid duplication

Reply:
@kamipo does `associated_primary_key` look better?